### ISSUE TITLE
segfault or ffi fatal: callback has been garbage collected when calling kernel too fast

### ISSUE THREAD

**Issue by vird** [unbornchikken/NOOOCL#9]
sorry for iced-coffee-script sources
https://gist.github.com/vird/9f041b18ad43ec9bfbfc

```
Error: ffi fatal: callback has been garbage collected!
  at /opt/ext/node_modules/nooocl/node_modules/ffi/lib/callback.js:18:15
  at process._tickCallback (node.js:419:13)
```

this also sometimes fire segfault

nooocl@0.9.10 node_modules/nooocl
├── bluebird@2.9.34
├── lodash@3.10.1
├── ref-struct@1.0.1 (debug@2.2.0)
├── ref-array@1.1.1 (array-index@0.1.1, debug@2.2.0)
├── weak@0.4.1 (bindings@1.2.1, nan@1.8.4)
└── ffi@1.3.2 (bindings@1.2.1, debug@2.2.0, nan@1.9.0)
node -v
v0.10.30
npm -v
1.4.21

**Comment by vird** [comment]
https://gist.github.com/vird/001de1a393adc9233047
reproduceable segfault

**Comment by vird** [comment]
seems 0.10.0 fixes ffi fatal: callback has been garbage collected, but I patched ffi-io -> ffi, so I don't know it's ok or not.

**Comment by unbornchikken** [comment]
Are you experiencing this by using 0.10.1?

**Comment by vird** [comment]
More minimalistic crash example with new nodejs.
Event without kernel call. Just memory IO.
https://gist.github.com/vird/2bb7e17ff34f3b24e069070d81cb1603
(iced-coffee-script)

    node -v
    v6.6.0
    cat node_modules/nooocl/package.json | grep version
    "version": "0.11.1"

P.s. you can remove enqueueReadBuffer or enqueueWriteBuffer, crash reproduces anyway.

**Comment by unbornchikken** [comment]
How do you spell this in ES6?

`await promise.then defer()`

**Comment by vird** [comment]
It's iced-coffee-script

**Comment by unbornchikken** [comment]
Well, I got it already. But why do you assume that I understand it? The problem is _thookh'aak see thuum ataa boon thaii_ and the workaround is obivous: _akhu'm wee khak shedo badaboom_. It's Klingon. :)

Sorry, just kidding. The problem is that you're creating a bunch of promises but those go out of scope before OpenCL event handler gets executed. This line somehow: `await promise.then defer()` is not defer the control flow until the promise resolves.

**Comment by vird** [comment]
Ok. Simpler example
https://gist.github.com/vird/fca26580b0456e768803ca1a315c21cb

**Comment by unbornchikken** [comment]
Fixed as of v0.12.1. I really am sorry for the delay, I was busy with [fastcall](https://github.com/cmake-js/fastcall) of which NOOOCL became to based on recently.
